server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java

```diff
@@ -168,21 +273,38 @@ public boolean equals(Object o) {
             && numCands == that.numCands
             && Objects.equals(field, that.field)
             && Arrays.equals(queryVector, that.queryVector)
+            && Objects.equals(queryVectorBuilder, that.queryVectorBuilder)
+            && Objects.equals(querySupplier, that.querySupplier)
             && Objects.equals(filterQueries, that.filterQueries)
             && boost == that.boost;
     }
 
     @Override
     public int hashCode() {
-        return Objects.hash(field, k, numCands, Arrays.hashCode(queryVector), Objects.hashCode(filterQueries), boost);
+        return Objects.hash(
+            field,
+            k,
+            numCands,
+            querySupplier,
+            queryVectorBuilder,
```

Comment:
would it make sense to somehow enforce that equals and hashcode are implemented on query vector builder? I worry that it can be easily forgotten.

Reply:
I can add them as base methods and thus requiring overriding.

Reply:
> would it make sense to somehow enforce that equals and hashcode are implemented on query vector builder? I worry that it can be easily forgotten.

Well, there is no simple way to do this with Java's class hierarchy. I would rather not make `QueryVectorBuilder` an abstract class. Tests should catch this fairly quickly though.

So, I am not going to enforce it in the class definition.